src/FreeAge/server/game.cpp

```diff
@@ -1349,21 +1401,137 @@ bool Game::SimulateMeleeAttack(u32 /*unitId*/, ServerUnit* unit, u32 targetId, S
   return true;
 }
 
-void Game::ProduceUnit(ServerBuilding* building, UnitType unitInProduction) {
+void Game::ProduceUnit(u32 buildingId, ServerBuilding* building, UnitType unitInProduction) {
   // Create the unit object.
   u32 newUnitId;
   ServerUnit* newUnit = map->AddUnit(building->GetPlayerIndex(), unitInProduction, QPointF(-999, -999), &newUnitId);
   
+  QPointF freeSpace;
+  bool foundFreeSpace = FindFreeSpaceAroundBuilding(building, newUnit, freeSpace);
+  
+  if (foundFreeSpace) {
+    newUnit->SetMapCoord(freeSpace);
+  } else {
+    // TODO: set a special value that will be ignored by later checks 
+    newUnit->SetMapCoord(building->GetBaseTile());
+  }
+  
+  // Send messages to clients that see the new unit
+  QByteArray addObjectMsg = CreateAddObjectMessage(newUnitId, newUnit);
+  for (auto& player : *playersInGame) {
+    accumulatedMessages[player->index] += addObjectMsg;
+  }
+
+  if (!foundFreeSpace) {
+    // TODO: what happens when the capacity is full ?
+    building->GarrisonUnit(newUnit);
+    newUnit->SetGarrisonedInsideObject(building);
+
+    QByteArray unitGarrisonMessage = CreateUnitGarrisonMessage(newUnitId, buildingId);
+    for (auto& player : *playersInGame) {
+      accumulatedMessages[player->index] += unitGarrisonMessage;
+    }
+  }
+  
+  GetPlayerStats(newUnit->GetPlayerIndex())->UnitAdded(unitInProduction);
+}
+
+void Game::GarrisonUnit(u32 unitId, ServerUnit* unit, u32 targetObjectId, ServerObject* targetObject, bool enter, bool* unitMovementChanged) {
+  // NOTE: Could be simplified if Objects stored there ID. #ids
+  *unitMovementChanged = false;
+  if (enter) {
+    // TODO: test with target object being a unit
+
+    unit->StopMovement();
+    unit->RemoveTarget();
+
+    // Check if the object is a building and has <= 20% hp left
+    if (targetObject->GetGarrisonedUnitsCount() > 0 && targetObject->isBuilding() &&
+        AsBuilding(targetObject)->GetHPInternalFloat() <= .2f * GetBuildingMaxHP(AsBuilding(targetObject)->GetType())) {
+      LOG(WARNING) << "Unit cannot garrison building beacause of low health";
```

Comment:
I think that this does not need to give a warning. Since the building may lose health between the garrison action and the unit reaching the building, it is normal that this happens.